src/Foundation/NSObject2.cs

```diff
@@ -155,6 +161,34 @@ internal static IntPtr CreateNSObject (IntPtr type_gchandle, IntPtr handle, Flag
 			return Runtime.AllocGCHandle (obj);
 		}
 
+		IntPtr GetSuper ()
+		{
+			if (super == IntPtr.Zero) {
+				IntPtr ptr;
+
+				unsafe {
+					ptr = Marshal.AllocHGlobal (sizeof (objc_super));
+					*(objc_super*) ptr = default (objc_super); // zero fill
+				}
+
+				var previousValue = Interlocked.CompareExchange (ref super, ptr, IntPtr.Zero);
+				if (previousValue != IntPtr.Zero) {
+					// somebody beat us to the assignment.
+					Marshal.FreeHGlobal (ptr);
+					ptr = IntPtr.Zero;
```

Comment:
minor: the local `ptr` is not used after `Marshal.FreeHGlobal` so it does not have to be set back to `IntPtr.Zero`

Reply:
I like it this way: the JIT/AOT compiler should be able to optimize the assignment away, and if the code is refactored on day it's easier to not forget that this pointer has been freed.